File: app/ios/WatchApp/Sources/TalkSession.swift
/// The state of one attempt to say something.
///
/// A press-and-hold reply is not instant the way a tap is: the clip has to reach the
/// phone and come back as text before there is anything to confirm. That gap is a few
/// seconds during which the user has already taken their eyes off the watch, so every
/// step of it is named and shown rather than hidden behind a spinner — if it fails,
/// they need to know whether to speak again or reach for the phone.
import Foundation
import Observation
import WatchKit

@Observable
@MainActor
final class TalkSession {
  static let shared = TalkSession()

  enum Phase: Equatable {
    case idle
    case recording
    case sending // clip on its way to the phone
    case transcribing // phone has it, recogniser running
    case failed(String)
  }

  private(set) var phase: Phase = .idle

  /// Correlates the clip we sent with the transcript that comes back.
  private(set) var clientId: String?

  /// Beyond this we stop waiting. The transfer is Bluetooth and the recogniser may
  /// hit the network, so it is generous — but not unbounded, because a spinner that
  /// never resolves is worse than an error that offers a way out.
  private static let transcriptTimeout: TimeInterval = 25

  private var watchdog: Task<Void, Never>?

  private init() {}

  var isBusy: Bool {
    switch phase {
    case .idle, .failed: return false
    case .recording, .sending, .transcribing: return true
    }
  }

  func beginRecording() {
    watchdog?.cancel()
    clientId = nil
    phase = .recording
  }

  /// The clip is captured; hand it over and start waiting for words.
  func submitted(clientId id: String) {
    clientId = id
    phase = .sending
    watchdog?.cancel()
    watchdog = Task { [weak self] in
      try? await Task.sleep(nanoseconds: UInt64(Self.transcriptTimeout * 1_000_000_000))
      guard !Task.isCancelled else { return }
      await MainActor.run { self?.fail("No reply from iPhone") }
    }
  }

  func phoneReceived() {
    guard case .sending = phase else { return }
    phase = .transcribing
  }

  /// Words the phone sent back, waiting for the confirm screen to pick them up.
  var pendingTranscript: String?

  /// How long the phone's recogniser took, and whether it ran on-device. Shown in
  /// Settings: a second on-device and eight over the network feel like the same
  /// unexplained wait from the wrist, and only one of them is worth doing anything
  /// about.
  private(set) var lastTranscribeMs: Int?
  private(set) var lastTranscribeOnDevice = false

  /// The phone's answer. Ignored unless it belongs to the attempt in flight — a
  /// slow transcript from an abandoned press must not hijack a later one.
  func deliver(clientId id: String, text: String?, error: String?,
               ms: Int? = nil, onDevice: Bool = false) {
    guard clientId == id else { return }
    watchdog?.cancel()
    clientId = nil
    if let ms { lastTranscribeMs = ms }
    lastTranscribeOnDevice = onDevice

    if let error {
      fail(error)
      return
    }
    let body = (text ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
    guard !body.isEmpty else {
      fail("Didn't catch that")
      return
    }
    phase = .idle
    pendingTranscript = body
  }

  func fail(_ reason: String) {
    watchdog?.cancel()
    phase = .failed(reason)
    clientId = nil
    WKInterfaceDevice.current().play(.failure)
  }

  func reset() {
    watchdog?.cancel()
    phase = .idle
    clientId = nil
  }

  var statusText: String? {
    switch phase {
    case .idle: return nil
    case .recording: return nil // the timer is shown instead
    case .sending: return "Sending audio…"
    case .transcribing: return "Transcribing…"
    case .failed(let why): return why
    }
  }
}

File: app/android/wear/src/main/kotlin/org/w6sg/aprsmap/wear/TalkSession.kt
/// The state of one attempt to say something.
///
/// Shorter-lived than its watchOS namesake, and deliberately so. There the clip has to
/// reach the phone and come back as text before there is anything to send, which is a
/// multi-second gap with three named stages and three ways to fail. Here the recogniser is
/// on the wrist (see SpeechCapture.kt), so the only wait is the recogniser finishing its
/// last word — but it is still a wait during which the operator has already taken their
/// eyes off the watch, so it is still named and shown rather than hidden behind a spinner.
///
/// Counterpart: `app/ios/WatchApp/Sources/TalkSession.swift`.
package org.w6sg.aprsmap.wear

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

object TalkSession {
    sealed interface Phase {
        object Idle : Phase
        object Listening : Phase
        object Transcribing : Phase
        data class Failed(val reason: String) : Phase
    }

    var phase by mutableStateOf<Phase>(Phase.Idle)
        private set

    /// Beyond this we stop waiting. Unbounded is worse than an error that offers a way out:
    /// a spinner that never resolves during a net is indistinguishable from a dead app.
    private const val RESULT_TIMEOUT_MS = 12_000L

    private var watchdog: Job? = null

    val isBusy: Boolean
        get() = phase is Phase.Listening || phase is Phase.Transcribing

    /// Words the recogniser produced, waiting for the Talk screen to pick them up and send.
    var pendingTranscript by mutableStateOf<String?>(null)

    fun beginListening() {
        watchdog?.cancel()
        phase = Phase.Listening
    }

    /// The finger is up and the recogniser is finishing. Start the clock: from here the app
    /// is waiting on something it does not control.
    fun transcribing() {
        if (phase !is Phase.Listening && phase !is Phase.Transcribing) return
        phase = Phase.Transcribing
        watchdog?.cancel()
        watchdog = mainScope.launch {
            delay(RESULT_TIMEOUT_MS)
            fail("No words came back")
        }
    }

    fun deliver(text: String) {
        watchdog?.cancel()
        val body = text.trim()
        if (body.isEmpty()) {
            fail("Didn't catch that")
            return
        }
        phase = Phase.Idle
        pendingTranscript = body
    }

    fun fail(reason: String) {
        watchdog?.cancel()
        phase = Phase.Failed(reason)
        Haptics.failure()
    }

    fun reset() {
        watchdog?.cancel()
        phase = Phase.Idle
    }

    /// What to put under the button. Null while listening — the timer is shown instead.
    val statusText: String?
        get() = when (val p = phase) {
            Phase.Idle -> null
            Phase.Listening -> null
            Phase.Transcribing -> "Transcribing…"
            is Phase.Failed -> p.reason
        }
}
